### PR TITLE
Add lazy loading of vscode snipets

### PR DESCRIPTION
**Pull request by leiserfg**
It can be improved once #58 get merged

### CODE REVIEW

**Comment by L3MON4D3** [comment]
Looks good so far :+1:

**Comment by leiserfg** [comment]
I noticed that sometimes the snippets get duplicated (seems like the function from @saadparwaiz1  can return the same path twice)  and my function to remove the duplicated is not working well, so I will remove the duplicated on a deeper level. Will update the pr when that gets fixed.

**Comment by leiserfg** [comment]
@L3MON4D3  have a look at it again please

**Comment by L3MON4D3** [comment]
> I noticed that sometimes the snippets get duplicated (seems like the function from @saadparwaiz1 can return the same path twice) and my function to remove the duplicated is not working well, so I will remove the duplicated on a deeper level. Will update the pr when that gets fixed.

 Are you certain that there are duplicate paths in the output of `get_snippets_rtp`? I'm a bit sceptical since each dir can only contain one `package.json`, therefore there shouldn't be duplicate paths?
I think not adding these would be nicer than removing after the fact.

**Comment by leiserfg** [comment]
It can, I don't know why, but any way doing it after the fact, allow us to remove other duplicated entries, like for instance, someone adding by hand a path that is already in the `runtimepath`.

**Comment by leiserfg** [comment]
I checked and the repetition comes from `vim.api.nvim_get_runtime_file()` itself.

**Comment by L3MON4D3** [comment]
Okay weird, wouldn't have thought that.
You're right, it won't hurt to make sure paths are unique.

**Comment by leiserfg** [comment]
> Okay weird, wouldn't have thought that.
> You're right, it won't hurt to make sure paths are unique.

Maybe is that packer is adding the plugin twice to the runtimepath  and `nvim_get_runtime_file` does not dedupe.

File: lua/luasnip/loaders/from_vscode.lua [hunk 1]
```diff
@@ -187,12 +187,42 @@ function M.load(opts)
 		opts.paths = get_snippets_rtp()
 	end
 
+	opts.paths = vim.tbl_map(expand_path, opts.paths) -- Expand before deduping, fake paths will become nil
```
Comment (L3MON4D3) [inline lua/luasnip/loaders/from_vscode.lua:190]: The paths from `get_snippets_rtp` don't need to be expanded anymore, right? In that case, I think this line should be inside the 'if'.

File: lua/luasnip/loaders/from_vscode.lua [hunk 2]
```diff
@@ -187,12 +187,42 @@ function M.load(opts)
...
 
+function M.lazy_load(opts)
+	opts = opts or {}
+
+	-- We have to do this here too, because we have to store them in lozy_load_paths
+	if type(opts.paths) ~= "table" and opts.paths ~= nil then
+		opts.paths = vim.split(opts.paths, ",")
+	else
+		opts.paths = get_snippets_rtp()
+	end
```
Comment (L3MON4D3) [inline lua/luasnip/loaders/from_vscode.lua:218]: The paths should also be filtered here, should they not?

File: lua/luasnip/loaders/from_vscode.lua (hunk 1 above)
Comment (leiserfg) [reply lua/luasnip/loaders/from_vscode.lua:190]: I'm not sure if it always brings the absolute path as `rutimepath` may contain paths with `~`, let me check.

File: lua/luasnip/loaders/from_vscode.lua [hunk 3]
```diff
@@ -187,12 +187,42 @@ function M.load(opts)
...
+	vim.list_extend(lazy_load_paths, opts.paths)
+	vim.cmd([[
+		augroup _luasnip_vscode_lazy_load
+		autocmd!
+		au BufEnter * lua _G._luasnip_vscode_lazy_load()
```
Comment (L3MON4D3) [inline lua/luasnip/loaders/from_vscode.lua:223]: tbh I'd rather not have `lazy_load` in the global namespace, the user probably won't interact with it directly. Maybe put the function in `M` and call here using `require`?

File: lua/luasnip/loaders/from_vscode.lua (hunk 2 above)
Comment (leiserfg) [reply lua/luasnip/loaders/from_vscode.lua:218]: Here is not that important, the only issue could be that the user calls the function too many times (like in a loop) cause that will make the cache of paths grow. But yes, let me at least dedupe them.

File: lua/luasnip/loaders/from_vscode.lua (hunk 3 above)
Comment (leiserfg) [reply lua/luasnip/loaders/from_vscode.lua:223]: Ok, I put it on _G as it's not exposed API, but if you prefer not to clutter the global context,  I will do it.

Comment (L3MON4D3) [reply lua/luasnip/loaders/from_vscode.lua:223]: ty!

File: lua/luasnip/loaders/from_vscode.lua (hunk 2 above)
Comment (L3MON4D3) [reply lua/luasnip/loaders/from_vscode.lua:218]: Ah ofc, completely overlooked that `M.load` is called on each path.

**Comment by L3MON4D3** [comment]
Nice, if you add the `M`-part I'm happy to merge :D

**Comment by leiserfg** [comment]
> Nice, if you add the `M._luasnip_vscode_lazy_load`-part I'm happy to merge :D

Sure, I thought you marked that as resolved, let me do it now.

**Comment by L3MON4D3** [comment]
Yeah, I thought you were going to add it in your next commit and marked it as resolved, but that's on me :sweat_smile:

**Comment by L3MON4D3** [comment]
Okay looks great, thanks a lot :D

**Pull request merged**